{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fed03c08-d374-4725-b73e-444569c138d8",
   "metadata": {},
   "source": [
    "# Getting Started with Alignment Files in Melodia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d62ae12c-eefd-4f10-83a2-025c9e3e0844",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dill\n",
    "import warnings\n",
    "\n",
    "import pandas as pd\n",
    "import melodia as mel\n",
    "import seaborn as sns\n",
    "\n",
    "from os import path\n",
    "from Bio.PDB.PDBExceptions import PDBConstructionWarning\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PDBConstructionWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2e343cc-13c5-4cbb-824c-4ce8347197b4",
   "metadata": {},
   "source": [
    "## Parsing an alignment in the PIR file format\n",
    "\n",
    "Melodia can read PIR alignment files and parse the geometric descriptors from the PDB files in the same directory.\n",
    "\n",
    "***\n",
    "![PIR record](model_pir.png)\n",
    "***\n",
    "\n",
    "In this example, the **structureX**  and the  **>P1;1cdoa** records inform Melodia that this sequence is related to a protein structure in a file called **1cdoa.pdb**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "70f59e69-cbe1-4d03-8e1e-f838472beb93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dill can be used for storage\n",
    "\n",
    "# Load the model if it already exists\n",
    "if path.exists('model.dill'):\n",
    "    with open('model.dill', 'rb') as file:\n",
    "        align = dill.load(file)\n",
    "else:\n",
    "    # Parser and save a new alignment\n",
    "    align = mel.parser_pir_file('model.ali')\n",
    "    with open('model.dill', 'wb') as file:\n",
    "        dill.dump(align, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be6e8318-15a4-490a-bf54-2917a1a22e27",
   "metadata": {},
   "source": [
    "The result is a BioPython alignment:\n",
    "\n",
    "https://biopython.org/docs/1.74/api/Bio.Align.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9a6f99ed-ac19-45ac-89cd-e119d67f7dfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<<class 'Bio.Align.MultipleSeqAlignment'> instance (6 records of length 375) at 7f73a8e27250>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e4f7e6e8-811c-4f90-8f83-ff6693816588",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: 1cdoa\n",
      "Name: 1cdoa\n",
      "Description: structureX:1cdoa:   1 :A: 374 :A:undefined:undefined::\n",
      "Number of features: 0\n",
      "/PIR-type=P1\n",
      "/molecule_type=protein\n",
      "Per letter annotation for: curvature, torsion, arc_length, writhing, phi, psi\n",
      "Seq('ATVGKVIKCKAAVAWEANKPLVIEEIEVDVPHANEIRIKIIATGVCHTDLYHLF...LSL')\n"
     ]
    }
   ],
   "source": [
    "# It easy to iterate over the alignment records \n",
    "for record in align:\n",
    "    print(record)\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c6c3e5d-e26b-4020-8a9a-788dcd1bfeb9",
   "metadata": {},
   "source": [
    "## Accessing Geometric Attributes\n",
    "\n",
    "All geometric attributes can be accessed through the letter_annotations funcionality:\n",
    "\n",
    "https://biopython.org/docs/1.75/api/Bio.SeqRecord.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6b0fe4ea-ca30-48e5-af17-e0654890fbb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1teha\n",
      "--ANEVIKCKAAVAWEAGKPLSIEEIEVAPPKAHEVRIKIIATAVCHTDAYTLSGADPEGCFPVILGHEGAGIVESVGEGVTKLKAGDTVIPLYIPQCGECKFCLNPKTNLCQKIRVTQGKGLMPDGTSRFTCKGKTILHYMGTSTFSEYTVVADISVAKIDPLAPLDKVCLLGCGISTGYGAAVNTAKLEPGSVCAVFGLGGVGLAVIMGCKVAGASRIIGVDINKDKFARAKEFGATECINPQDFSKPIQEVLIEMTDGGVDYSFECIGNVKVMRAALEACHKGWGVSVVVGVAASGEEIATRPFQLVTGRTWKGTAFGGWKSVESVPKLVSEYMSKKIKVDEFVTHNLSFDEINKAFELMHSGKSIRTVVKI\n",
      "dict_keys(['curvature', 'torsion', 'arc_length', 'writhing', 'phi', 'psi'])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Select the third sequence in the alignment\n",
    "record = align[2]\n",
    "\n",
    "# Print some of the record's data\n",
    "print(record.id)\n",
    "print(record.seq)\n",
    "print(record.letter_annotations.keys())\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63483f37-bd27-426e-a8c8-8dcd2e5e355f",
   "metadata": {},
   "source": [
    "It is easy to access and work with the alignment data with the geometric annotation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "92a941ab-ff3f-4d55-aeed-7b4d5ce9e175",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 - - -  0.0000 -  0.0000\n",
      "1 - - -  0.0000 -  0.0000\n",
      "2 - A -  0.6445 -  0.1214\n",
      "3 - N -  0.6445 -  0.1214\n",
      "4 - E -  0.5355 - -0.0036\n",
      "5 - V -  0.9706 - -0.1388\n"
     ]
    }
   ],
   "source": [
    "# Print the curvature and torsion for a few residues\n",
    "for i, residue in enumerate(record.seq):\n",
    "    print(f\"{i} - {residue} - {record.letter_annotations['curvature'][i]:7.4f} - {record.letter_annotations['torsion'][i]:7.4f}\")\n",
    "    if i > 4:\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d12e2911-530c-4015-8994-fd6c89ca2591",
   "metadata": {},
   "source": [
    "## Alignments and Pandas Dataframe\n",
    "\n",
    "### Converting a BioPython alignment to a Pandas Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d186d4ba-945f-4c74-9eb8-696925183224",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_1cdoa</th>\n",
       "      <th>curvature_1cdoa</th>\n",
       "      <th>torsion_1cdoa</th>\n",
       "      <th>arc_length_1cdoa</th>\n",
       "      <th>writhing_1cdoa</th>\n",
       "      <th>phi_1cdoa</th>\n",
       "      <th>psi_1cdoa</th>\n",
       "      <th>seq_1d1ta</th>\n",
       "      <th>curvature_1d1ta</th>\n",
       "      <th>torsion_1d1ta</th>\n",
       "      <th>...</th>\n",
       "      <th>writhing_2ohxa</th>\n",
       "      <th>phi_2ohxa</th>\n",
       "      <th>psi_2ohxa</th>\n",
       "      <th>seq_3huda</th>\n",
       "      <th>curvature_3huda</th>\n",
       "      <th>torsion_3huda</th>\n",
       "      <th>arc_length_3huda</th>\n",
       "      <th>writhing_3huda</th>\n",
       "      <th>phi_3huda</th>\n",
       "      <th>psi_3huda</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>0.911238</td>\n",
       "      <td>0.110831</td>\n",
       "      <td>8.246699</td>\n",
       "      <td>0.084959</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-10.148714</td>\n",
       "      <td>G</td>\n",
       "      <td>1.047306</td>\n",
       "      <td>0.110002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.084033</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>164.623349</td>\n",
       "      <td>S</td>\n",
       "      <td>0.794220</td>\n",
       "      <td>0.107699</td>\n",
       "      <td>8.312687</td>\n",
       "      <td>0.090238</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-106.741610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>0.911238</td>\n",
       "      <td>0.110831</td>\n",
       "      <td>7.886044</td>\n",
       "      <td>0.084959</td>\n",
       "      <td>-116.979109</td>\n",
       "      <td>-17.230205</td>\n",
       "      <td>T</td>\n",
       "      <td>1.047306</td>\n",
       "      <td>0.110002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.084033</td>\n",
       "      <td>-111.321155</td>\n",
       "      <td>-15.961734</td>\n",
       "      <td>T</td>\n",
       "      <td>0.794220</td>\n",
       "      <td>0.107699</td>\n",
       "      <td>7.870623</td>\n",
       "      <td>0.090238</td>\n",
       "      <td>-109.782193</td>\n",
       "      <td>-30.437007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V</td>\n",
       "      <td>0.434979</td>\n",
       "      <td>0.044909</td>\n",
       "      <td>8.043852</td>\n",
       "      <td>0.084959</td>\n",
       "      <td>-55.693954</td>\n",
       "      <td>132.988680</td>\n",
       "      <td>A</td>\n",
       "      <td>0.454057</td>\n",
       "      <td>0.032614</td>\n",
       "      <td>...</td>\n",
       "      <td>0.084033</td>\n",
       "      <td>-52.657511</td>\n",
       "      <td>128.383430</td>\n",
       "      <td>A</td>\n",
       "      <td>0.463674</td>\n",
       "      <td>0.074985</td>\n",
       "      <td>7.973132</td>\n",
       "      <td>0.090238</td>\n",
       "      <td>-40.676160</td>\n",
       "      <td>125.848618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>0.822893</td>\n",
       "      <td>0.062217</td>\n",
       "      <td>8.030083</td>\n",
       "      <td>0.092038</td>\n",
       "      <td>78.090579</td>\n",
       "      <td>4.118510</td>\n",
       "      <td>G</td>\n",
       "      <td>0.801947</td>\n",
       "      <td>0.048148</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083461</td>\n",
       "      <td>88.115088</td>\n",
       "      <td>-16.629985</td>\n",
       "      <td>G</td>\n",
       "      <td>0.572861</td>\n",
       "      <td>0.099983</td>\n",
       "      <td>8.011114</td>\n",
       "      <td>0.120894</td>\n",
       "      <td>113.767325</td>\n",
       "      <td>-43.875177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K</td>\n",
       "      <td>0.390131</td>\n",
       "      <td>-0.000838</td>\n",
       "      <td>8.154757</td>\n",
       "      <td>0.010612</td>\n",
       "      <td>-123.849409</td>\n",
       "      <td>155.744956</td>\n",
       "      <td>K</td>\n",
       "      <td>0.504131</td>\n",
       "      <td>-0.012812</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006618</td>\n",
       "      <td>-110.955311</td>\n",
       "      <td>167.680235</td>\n",
       "      <td>K</td>\n",
       "      <td>0.347366</td>\n",
       "      <td>0.004505</td>\n",
       "      <td>8.178899</td>\n",
       "      <td>0.010359</td>\n",
       "      <td>-103.640973</td>\n",
       "      <td>177.954331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>T</td>\n",
       "      <td>0.061095</td>\n",
       "      <td>-0.749861</td>\n",
       "      <td>7.801091</td>\n",
       "      <td>-0.019713</td>\n",
       "      <td>-107.281588</td>\n",
       "      <td>129.499684</td>\n",
       "      <td>T</td>\n",
       "      <td>0.034498</td>\n",
       "      <td>-0.396657</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.019491</td>\n",
       "      <td>-103.494266</td>\n",
       "      <td>131.348977</td>\n",
       "      <td>T</td>\n",
       "      <td>0.054648</td>\n",
       "      <td>-0.794151</td>\n",
       "      <td>7.924285</td>\n",
       "      <td>-0.023005</td>\n",
       "      <td>-99.772840</td>\n",
       "      <td>127.501598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>V</td>\n",
       "      <td>0.023665</td>\n",
       "      <td>2.478917</td>\n",
       "      <td>7.834701</td>\n",
       "      <td>-0.021218</td>\n",
       "      <td>-116.081018</td>\n",
       "      <td>123.375950</td>\n",
       "      <td>V</td>\n",
       "      <td>0.044883</td>\n",
       "      <td>1.333120</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.025700</td>\n",
       "      <td>-102.055498</td>\n",
       "      <td>134.368430</td>\n",
       "      <td>V</td>\n",
       "      <td>0.034529</td>\n",
       "      <td>1.895340</td>\n",
       "      <td>7.871723</td>\n",
       "      <td>-0.022803</td>\n",
       "      <td>-109.796765</td>\n",
       "      <td>135.339999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>L</td>\n",
       "      <td>0.120265</td>\n",
       "      <td>0.082923</td>\n",
       "      <td>7.817057</td>\n",
       "      <td>-0.019700</td>\n",
       "      <td>-90.226598</td>\n",
       "      <td>143.795825</td>\n",
       "      <td>L</td>\n",
       "      <td>0.149571</td>\n",
       "      <td>0.055948</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.024220</td>\n",
       "      <td>-105.139660</td>\n",
       "      <td>133.371850</td>\n",
       "      <td>L</td>\n",
       "      <td>0.144089</td>\n",
       "      <td>0.042052</td>\n",
       "      <td>7.732766</td>\n",
       "      <td>-0.023656</td>\n",
       "      <td>-99.507560</td>\n",
       "      <td>126.540662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>S</td>\n",
       "      <td>0.176591</td>\n",
       "      <td>0.207346</td>\n",
       "      <td>7.868773</td>\n",
       "      <td>-0.019700</td>\n",
       "      <td>-100.424779</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>0.098639</td>\n",
       "      <td>-0.130310</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.024220</td>\n",
       "      <td>-109.600320</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>0.139156</td>\n",
       "      <td>-0.010417</td>\n",
       "      <td>7.734215</td>\n",
       "      <td>-0.023656</td>\n",
       "      <td>-109.028007</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>L</td>\n",
       "      <td>0.176591</td>\n",
       "      <td>0.207346</td>\n",
       "      <td>7.966369</td>\n",
       "      <td>-0.019700</td>\n",
       "      <td>-134.654331</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>F</td>\n",
       "      <td>0.098639</td>\n",
       "      <td>-0.130310</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.024220</td>\n",
       "      <td>-100.989217</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>F</td>\n",
       "      <td>0.139156</td>\n",
       "      <td>-0.010417</td>\n",
       "      <td>7.771344</td>\n",
       "      <td>-0.023656</td>\n",
       "      <td>-102.480153</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>375 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seq_1cdoa  curvature_1cdoa  torsion_1cdoa  arc_length_1cdoa  \\\n",
       "0           A         0.911238       0.110831          8.246699   \n",
       "1           T         0.911238       0.110831          7.886044   \n",
       "2           V         0.434979       0.044909          8.043852   \n",
       "3           G         0.822893       0.062217          8.030083   \n",
       "4           K         0.390131      -0.000838          8.154757   \n",
       "..        ...              ...            ...               ...   \n",
       "370         T         0.061095      -0.749861          7.801091   \n",
       "371         V         0.023665       2.478917          7.834701   \n",
       "372         L         0.120265       0.082923          7.817057   \n",
       "373         S         0.176591       0.207346          7.868773   \n",
       "374         L         0.176591       0.207346          7.966369   \n",
       "\n",
       "     writhing_1cdoa   phi_1cdoa   psi_1cdoa seq_1d1ta  curvature_1d1ta  \\\n",
       "0          0.084959    0.000000  -10.148714         G         1.047306   \n",
       "1          0.084959 -116.979109  -17.230205         T         1.047306   \n",
       "2          0.084959  -55.693954  132.988680         A         0.454057   \n",
       "3          0.092038   78.090579    4.118510         G         0.801947   \n",
       "4          0.010612 -123.849409  155.744956         K         0.504131   \n",
       "..              ...         ...         ...       ...              ...   \n",
       "370       -0.019713 -107.281588  129.499684         T         0.034498   \n",
       "371       -0.021218 -116.081018  123.375950         V         0.044883   \n",
       "372       -0.019700  -90.226598  143.795825         L         0.149571   \n",
       "373       -0.019700 -100.424779    0.000000         T         0.098639   \n",
       "374       -0.019700 -134.654331    0.000000         F         0.098639   \n",
       "\n",
       "     torsion_1d1ta  ...  writhing_2ohxa   phi_2ohxa   psi_2ohxa  seq_3huda  \\\n",
       "0         0.110002  ...        0.084033    0.000000  164.623349          S   \n",
       "1         0.110002  ...        0.084033 -111.321155  -15.961734          T   \n",
       "2         0.032614  ...        0.084033  -52.657511  128.383430          A   \n",
       "3         0.048148  ...        0.083461   88.115088  -16.629985          G   \n",
       "4        -0.012812  ...        0.006618 -110.955311  167.680235          K   \n",
       "..             ...  ...             ...         ...         ...        ...   \n",
       "370      -0.396657  ...       -0.019491 -103.494266  131.348977          T   \n",
       "371       1.333120  ...       -0.025700 -102.055498  134.368430          V   \n",
       "372       0.055948  ...       -0.024220 -105.139660  133.371850          L   \n",
       "373      -0.130310  ...       -0.024220 -109.600320    0.000000          T   \n",
       "374      -0.130310  ...       -0.024220 -100.989217    0.000000          F   \n",
       "\n",
       "    curvature_3huda  torsion_3huda  arc_length_3huda  writhing_3huda  \\\n",
       "0          0.794220       0.107699          8.312687        0.090238   \n",
       "1          0.794220       0.107699          7.870623        0.090238   \n",
       "2          0.463674       0.074985          7.973132        0.090238   \n",
       "3          0.572861       0.099983          8.011114        0.120894   \n",
       "4          0.347366       0.004505          8.178899        0.010359   \n",
       "..              ...            ...               ...             ...   \n",
       "370        0.054648      -0.794151          7.924285       -0.023005   \n",
       "371        0.034529       1.895340          7.871723       -0.022803   \n",
       "372        0.144089       0.042052          7.732766       -0.023656   \n",
       "373        0.139156      -0.010417          7.734215       -0.023656   \n",
       "374        0.139156      -0.010417          7.771344       -0.023656   \n",
       "\n",
       "      phi_3huda   psi_3huda  \n",
       "0      0.000000 -106.741610  \n",
       "1   -109.782193  -30.437007  \n",
       "2    -40.676160  125.848618  \n",
       "3    113.767325  -43.875177  \n",
       "4   -103.640973  177.954331  \n",
       "..          ...         ...  \n",
       "370  -99.772840  127.501598  \n",
       "371 -109.796765  135.339999  \n",
       "372  -99.507560  126.540662  \n",
       "373 -109.028007    0.000000  \n",
       "374 -102.480153    0.000000  \n",
       "\n",
       "[375 rows x 35 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mel.dataframe_from_alignment(align=align)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fba4d38c-c08a-4c99-b0de-8e3ce1d524ce",
   "metadata": {},
   "source": [
    "***\n",
    "It is possible to choose the geometric annotations for the Dataframe.\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "62587060-fdca-477e-bcf1-e381490dcc0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = mel.dataframe_from_alignment(align=align, keys=['curvature', 'torsion'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1e18a1b6-c1e4-4f5e-bafc-3b852859fee8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_1cdoa</th>\n",
       "      <th>curvature_1cdoa</th>\n",
       "      <th>torsion_1cdoa</th>\n",
       "      <th>seq_1d1ta</th>\n",
       "      <th>curvature_1d1ta</th>\n",
       "      <th>torsion_1d1ta</th>\n",
       "      <th>seq_1teha</th>\n",
       "      <th>curvature_1teha</th>\n",
       "      <th>torsion_1teha</th>\n",
       "      <th>seq_2ohxa</th>\n",
       "      <th>curvature_2ohxa</th>\n",
       "      <th>torsion_2ohxa</th>\n",
       "      <th>seq_3huda</th>\n",
       "      <th>curvature_3huda</th>\n",
       "      <th>torsion_3huda</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>0.911238</td>\n",
       "      <td>0.110831</td>\n",
       "      <td>G</td>\n",
       "      <td>1.047306</td>\n",
       "      <td>0.110002</td>\n",
       "      <td>-</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>S</td>\n",
       "      <td>0.858952</td>\n",
       "      <td>0.103759</td>\n",
       "      <td>S</td>\n",
       "      <td>0.794220</td>\n",
       "      <td>0.107699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>0.911238</td>\n",
       "      <td>0.110831</td>\n",
       "      <td>T</td>\n",
       "      <td>1.047306</td>\n",
       "      <td>0.110002</td>\n",
       "      <td>-</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>0.858952</td>\n",
       "      <td>0.103759</td>\n",
       "      <td>T</td>\n",
       "      <td>0.794220</td>\n",
       "      <td>0.107699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V</td>\n",
       "      <td>0.434979</td>\n",
       "      <td>0.044909</td>\n",
       "      <td>A</td>\n",
       "      <td>0.454057</td>\n",
       "      <td>0.032614</td>\n",
       "      <td>A</td>\n",
       "      <td>0.644497</td>\n",
       "      <td>0.121356</td>\n",
       "      <td>A</td>\n",
       "      <td>0.505272</td>\n",
       "      <td>0.043087</td>\n",
       "      <td>A</td>\n",
       "      <td>0.463674</td>\n",
       "      <td>0.074985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>0.822893</td>\n",
       "      <td>0.062217</td>\n",
       "      <td>G</td>\n",
       "      <td>0.801947</td>\n",
       "      <td>0.048148</td>\n",
       "      <td>N</td>\n",
       "      <td>0.644497</td>\n",
       "      <td>0.121356</td>\n",
       "      <td>G</td>\n",
       "      <td>0.697603</td>\n",
       "      <td>0.059488</td>\n",
       "      <td>G</td>\n",
       "      <td>0.572861</td>\n",
       "      <td>0.099983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K</td>\n",
       "      <td>0.390131</td>\n",
       "      <td>-0.000838</td>\n",
       "      <td>K</td>\n",
       "      <td>0.504131</td>\n",
       "      <td>-0.012812</td>\n",
       "      <td>E</td>\n",
       "      <td>0.535522</td>\n",
       "      <td>-0.003564</td>\n",
       "      <td>K</td>\n",
       "      <td>0.366238</td>\n",
       "      <td>-0.007581</td>\n",
       "      <td>K</td>\n",
       "      <td>0.347366</td>\n",
       "      <td>0.004505</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seq_1cdoa  curvature_1cdoa  torsion_1cdoa seq_1d1ta  curvature_1d1ta  \\\n",
       "0         A         0.911238       0.110831         G         1.047306   \n",
       "1         T         0.911238       0.110831         T         1.047306   \n",
       "2         V         0.434979       0.044909         A         0.454057   \n",
       "3         G         0.822893       0.062217         G         0.801947   \n",
       "4         K         0.390131      -0.000838         K         0.504131   \n",
       "\n",
       "   torsion_1d1ta seq_1teha  curvature_1teha  torsion_1teha seq_2ohxa  \\\n",
       "0       0.110002         -         0.000000       0.000000         S   \n",
       "1       0.110002         -         0.000000       0.000000         T   \n",
       "2       0.032614         A         0.644497       0.121356         A   \n",
       "3       0.048148         N         0.644497       0.121356         G   \n",
       "4      -0.012812         E         0.535522      -0.003564         K   \n",
       "\n",
       "   curvature_2ohxa  torsion_2ohxa seq_3huda  curvature_3huda  torsion_3huda  \n",
       "0         0.858952       0.103759         S         0.794220       0.107699  \n",
       "1         0.858952       0.103759         T         0.794220       0.107699  \n",
       "2         0.505272       0.043087         A         0.463674       0.074985  \n",
       "3         0.697603       0.059488         G         0.572861       0.099983  \n",
       "4         0.366238      -0.007581         K         0.347366       0.004505  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bce9136c-ef30-4fbb-84fa-45e90587fa8a",
   "metadata": {},
   "source": [
    "### DataFrame Storage"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acbe8a61-f6b7-4281-b12b-abd06f1d9c0c",
   "metadata": {},
   "source": [
    "***\n",
    "Pandas Dataframe can be stored using the Parquet file format.\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8f591e7b-cf6e-46db-824c-6d4472534906",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_parquet('df.parquet.gzip', compression='gzip')  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a297d32c-a9c4-4fed-aa9e-5cea5a0f9759",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_1cdoa</th>\n",
       "      <th>curvature_1cdoa</th>\n",
       "      <th>torsion_1cdoa</th>\n",
       "      <th>seq_1d1ta</th>\n",
       "      <th>curvature_1d1ta</th>\n",
       "      <th>torsion_1d1ta</th>\n",
       "      <th>seq_1teha</th>\n",
       "      <th>curvature_1teha</th>\n",
       "      <th>torsion_1teha</th>\n",
       "      <th>seq_2ohxa</th>\n",
       "      <th>curvature_2ohxa</th>\n",
       "      <th>torsion_2ohxa</th>\n",
       "      <th>seq_3huda</th>\n",
       "      <th>curvature_3huda</th>\n",
       "      <th>torsion_3huda</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>0.911238</td>\n",
       "      <td>0.110831</td>\n",
       "      <td>G</td>\n",
       "      <td>1.047306</td>\n",
       "      <td>0.110002</td>\n",
       "      <td>-</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>S</td>\n",
       "      <td>0.858952</td>\n",
       "      <td>0.103759</td>\n",
       "      <td>S</td>\n",
       "      <td>0.794220</td>\n",
       "      <td>0.107699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>0.911238</td>\n",
       "      <td>0.110831</td>\n",
       "      <td>T</td>\n",
       "      <td>1.047306</td>\n",
       "      <td>0.110002</td>\n",
       "      <td>-</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>0.858952</td>\n",
       "      <td>0.103759</td>\n",
       "      <td>T</td>\n",
       "      <td>0.794220</td>\n",
       "      <td>0.107699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V</td>\n",
       "      <td>0.434979</td>\n",
       "      <td>0.044909</td>\n",
       "      <td>A</td>\n",
       "      <td>0.454057</td>\n",
       "      <td>0.032614</td>\n",
       "      <td>A</td>\n",
       "      <td>0.644497</td>\n",
       "      <td>0.121356</td>\n",
       "      <td>A</td>\n",
       "      <td>0.505272</td>\n",
       "      <td>0.043087</td>\n",
       "      <td>A</td>\n",
       "      <td>0.463674</td>\n",
       "      <td>0.074985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>0.822893</td>\n",
       "      <td>0.062217</td>\n",
       "      <td>G</td>\n",
       "      <td>0.801947</td>\n",
       "      <td>0.048148</td>\n",
       "      <td>N</td>\n",
       "      <td>0.644497</td>\n",
       "      <td>0.121356</td>\n",
       "      <td>G</td>\n",
       "      <td>0.697603</td>\n",
       "      <td>0.059488</td>\n",
       "      <td>G</td>\n",
       "      <td>0.572861</td>\n",
       "      <td>0.099983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K</td>\n",
       "      <td>0.390131</td>\n",
       "      <td>-0.000838</td>\n",
       "      <td>K</td>\n",
       "      <td>0.504131</td>\n",
       "      <td>-0.012812</td>\n",
       "      <td>E</td>\n",
       "      <td>0.535522</td>\n",
       "      <td>-0.003564</td>\n",
       "      <td>K</td>\n",
       "      <td>0.366238</td>\n",
       "      <td>-0.007581</td>\n",
       "      <td>K</td>\n",
       "      <td>0.347366</td>\n",
       "      <td>0.004505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>T</td>\n",
       "      <td>0.061095</td>\n",
       "      <td>-0.749861</td>\n",
       "      <td>T</td>\n",
       "      <td>0.034498</td>\n",
       "      <td>-0.396657</td>\n",
       "      <td>T</td>\n",
       "      <td>0.064315</td>\n",
       "      <td>-0.835171</td>\n",
       "      <td>T</td>\n",
       "      <td>0.045135</td>\n",
       "      <td>-0.640672</td>\n",
       "      <td>T</td>\n",
       "      <td>0.054648</td>\n",
       "      <td>-0.794151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>V</td>\n",
       "      <td>0.023665</td>\n",
       "      <td>2.478917</td>\n",
       "      <td>V</td>\n",
       "      <td>0.044883</td>\n",
       "      <td>1.333120</td>\n",
       "      <td>V</td>\n",
       "      <td>0.038668</td>\n",
       "      <td>1.412671</td>\n",
       "      <td>I</td>\n",
       "      <td>0.027378</td>\n",
       "      <td>2.689518</td>\n",
       "      <td>V</td>\n",
       "      <td>0.034529</td>\n",
       "      <td>1.895340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>L</td>\n",
       "      <td>0.120265</td>\n",
       "      <td>0.082923</td>\n",
       "      <td>L</td>\n",
       "      <td>0.149571</td>\n",
       "      <td>0.055948</td>\n",
       "      <td>V</td>\n",
       "      <td>0.084575</td>\n",
       "      <td>0.152802</td>\n",
       "      <td>L</td>\n",
       "      <td>0.150112</td>\n",
       "      <td>0.031978</td>\n",
       "      <td>L</td>\n",
       "      <td>0.144089</td>\n",
       "      <td>0.042052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>S</td>\n",
       "      <td>0.176591</td>\n",
       "      <td>0.207346</td>\n",
       "      <td>T</td>\n",
       "      <td>0.098639</td>\n",
       "      <td>-0.130310</td>\n",
       "      <td>K</td>\n",
       "      <td>0.248368</td>\n",
       "      <td>0.170128</td>\n",
       "      <td>T</td>\n",
       "      <td>0.125465</td>\n",
       "      <td>-0.016873</td>\n",
       "      <td>T</td>\n",
       "      <td>0.139156</td>\n",
       "      <td>-0.010417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>L</td>\n",
       "      <td>0.176591</td>\n",
       "      <td>0.207346</td>\n",
       "      <td>F</td>\n",
       "      <td>0.098639</td>\n",
       "      <td>-0.130310</td>\n",
       "      <td>I</td>\n",
       "      <td>0.248368</td>\n",
       "      <td>0.170128</td>\n",
       "      <td>F</td>\n",
       "      <td>0.125465</td>\n",
       "      <td>-0.016873</td>\n",
       "      <td>F</td>\n",
       "      <td>0.139156</td>\n",
       "      <td>-0.010417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>375 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seq_1cdoa  curvature_1cdoa  torsion_1cdoa seq_1d1ta  curvature_1d1ta  \\\n",
       "0           A         0.911238       0.110831         G         1.047306   \n",
       "1           T         0.911238       0.110831         T         1.047306   \n",
       "2           V         0.434979       0.044909         A         0.454057   \n",
       "3           G         0.822893       0.062217         G         0.801947   \n",
       "4           K         0.390131      -0.000838         K         0.504131   \n",
       "..        ...              ...            ...       ...              ...   \n",
       "370         T         0.061095      -0.749861         T         0.034498   \n",
       "371         V         0.023665       2.478917         V         0.044883   \n",
       "372         L         0.120265       0.082923         L         0.149571   \n",
       "373         S         0.176591       0.207346         T         0.098639   \n",
       "374         L         0.176591       0.207346         F         0.098639   \n",
       "\n",
       "     torsion_1d1ta seq_1teha  curvature_1teha  torsion_1teha seq_2ohxa  \\\n",
       "0         0.110002         -         0.000000       0.000000         S   \n",
       "1         0.110002         -         0.000000       0.000000         T   \n",
       "2         0.032614         A         0.644497       0.121356         A   \n",
       "3         0.048148         N         0.644497       0.121356         G   \n",
       "4        -0.012812         E         0.535522      -0.003564         K   \n",
       "..             ...       ...              ...            ...       ...   \n",
       "370      -0.396657         T         0.064315      -0.835171         T   \n",
       "371       1.333120         V         0.038668       1.412671         I   \n",
       "372       0.055948         V         0.084575       0.152802         L   \n",
       "373      -0.130310         K         0.248368       0.170128         T   \n",
       "374      -0.130310         I         0.248368       0.170128         F   \n",
       "\n",
       "     curvature_2ohxa  torsion_2ohxa seq_3huda  curvature_3huda  torsion_3huda  \n",
       "0           0.858952       0.103759         S         0.794220       0.107699  \n",
       "1           0.858952       0.103759         T         0.794220       0.107699  \n",
       "2           0.505272       0.043087         A         0.463674       0.074985  \n",
       "3           0.697603       0.059488         G         0.572861       0.099983  \n",
       "4           0.366238      -0.007581         K         0.347366       0.004505  \n",
       "..               ...            ...       ...              ...            ...  \n",
       "370         0.045135      -0.640672         T         0.054648      -0.794151  \n",
       "371         0.027378       2.689518         V         0.034529       1.895340  \n",
       "372         0.150112       0.031978         L         0.144089       0.042052  \n",
       "373         0.125465      -0.016873         T         0.139156      -0.010417  \n",
       "374         0.125465      -0.016873         F         0.139156      -0.010417  \n",
       "\n",
       "[375 rows x 15 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_parquet('df.parquet.gzip') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "193560f7-4368-4208-bdc8-0d41926f2a6b",
   "metadata": {},
   "source": [
    "## Structural Similarity Analysis\n",
    "\n",
    "### Alignment Clustering\n",
    "\n",
    "Melodia can cluster segments of the proteins in the alignment to determine the highly conserved regions. It infers the conservation patterns by comparing the differential geometry of aligned positions. Melodia uses a threshold to determine and group the regions where curvature and torsion are deemed similar. \n",
    "\n",
    "For more information about structural clustering see:\n",
    "\n",
    "Rinaldo W. Montalvão, Richard E. Smith, Simon C. Lovell, Tom L. Blundell, CHORAL: a differential geometry approach to the prediction of the cores of protein structures, *Bioinformatics*, Volume 21, Issue 19, January 2005, Pages 3719–3725, https://doi.org/10.1093/bioinformatics/bti595\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d1e78750-1fe5-49bd-abdc-1da7a13c7808",
   "metadata": {},
   "outputs": [],
   "source": [
    "mel.cluster_alignment(align=align, threshold=1.1, long=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ff72009-8e80-449e-a791-3f45faf038fe",
   "metadata": {},
   "source": [
    "### Alignment Cluster Annotation\n",
    "\n",
    "Melodia can save a colour-annotated version of the alignment as a PostScript file. Each colour, in aligned positions, indicates blocks with similar geometry. Those clusters accommodate the multimodal nature of protein ensembles of a homologous family, which is particularly characteristic of the Cα spatial distributions in low similarity superfamilies and for loop regions. The differential geometric classification provides a better classification than that obtained using Cα distances alone."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cb09bab2-cbc6-4509-931a-b7031f7b1a2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<svg  width=\"385\" height=\"55\"><rect x=\"0\" y=\"0\" width=\"55\" height=\"55\" style=\"fill:#1b9e77;stroke-width:2;stroke:rgb(255,255,255)\"/><rect x=\"55\" y=\"0\" width=\"55\" height=\"55\" style=\"fill:#d95f02;stroke-width:2;stroke:rgb(255,255,255)\"/><rect x=\"110\" y=\"0\" width=\"55\" height=\"55\" style=\"fill:#7570b3;stroke-width:2;stroke:rgb(255,255,255)\"/><rect x=\"165\" y=\"0\" width=\"55\" height=\"55\" style=\"fill:#e7298a;stroke-width:2;stroke:rgb(255,255,255)\"/><rect x=\"220\" y=\"0\" width=\"55\" height=\"55\" style=\"fill:#66a61e;stroke-width:2;stroke:rgb(255,255,255)\"/><rect x=\"275\" y=\"0\" width=\"55\" height=\"55\" style=\"fill:#e6ab02;stroke-width:2;stroke:rgb(255,255,255)\"/><rect x=\"330\" y=\"0\" width=\"55\" height=\"55\" style=\"fill:#a6761d;stroke-width:2;stroke:rgb(255,255,255)\"/></svg>"
      ],
      "text/plain": [
       "[(0.10588235294117647, 0.6196078431372549, 0.4666666666666667),\n",
       " (0.8509803921568627, 0.37254901960784315, 0.00784313725490196),\n",
       " (0.4588235294117647, 0.4392156862745098, 0.7019607843137254),\n",
       " (0.9058823529411765, 0.1607843137254902, 0.5411764705882353),\n",
       " (0.4, 0.6509803921568628, 0.11764705882352941),\n",
       " (0.9019607843137255, 0.6705882352941176, 0.00784313725490196),\n",
       " (0.6509803921568628, 0.4627450980392157, 0.11372549019607843)]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First select a colour pallete\n",
    "palette='Dark2'\n",
    "colors=7\n",
    "sns.color_palette(palette, colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "210602de-3ee8-4c3c-abb4-9be41b24a315",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save a PS file with the colour-coded alignment\n",
    "mel.save_align_to_ps(align=align, ps_file='model', palette=palette, colors=colors)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4abb784-68a4-4ae5-ae94-2772a391fcc4",
   "metadata": {},
   "source": [
    "![alignment cluster](ali_cluster.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9df5840-a6f0-47bb-8e78-3e27fe2a2a20",
   "metadata": {},
   "source": [
    "### Structure Superpostion and Annotation\n",
    "\n",
    "Melodia can also create a PyMol script to load and superpose the protein structures and colour the clustered regions like in the PS file. The following command will produce a **clusters_model.pml** script file for this operation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "07a76f8e-268c-46cb-865c-900d71a38e60",
   "metadata": {},
   "outputs": [],
   "source": [
    "mel.save_pymol_script(align=align, pml_file='cluster_models', palette=palette, colors=colors)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f670d788-d3ce-4a98-986f-88a4c4fe2c9e",
   "metadata": {},
   "source": [
    "Just run **pymol cluster_models.pml** in the command line to create this vizualization.\n",
    "\n",
    "![pdb clusters](pdbs_cluster.png)\n",
    "\n",
    "***\n",
    "Using a colour-coded alignment and colour-coded structures can give valuable insights into the structural conservation patterns of a protein family. Colour-coded structures can indicate different conformational patterns in an otherwise reasonably similar region that would be classified as a common framework in many programs.\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b3fa2ba-84f1-456e-86b1-e6cf1441f246",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
